freud/common.pyx

```diff
@@ -0,0 +1,179 @@
+# Copyright (c) 2010-2019 The Regents of the University of Michigan
+# This file is from the freud project, released under the BSD 3-Clause License.
+
+# Methods used throughout freud for convenience
+
+import logging
+import numpy as np
+import freud.box
+
+from functools import wraps
+
+logger = logging.getLogger(__name__)
+
+
+cdef class Compute:
+    R"""Parent class implementing functions to prevent access of
+    uncomputed values.
+
+    To use this class, one would do, for example,
+
+    .. code-block:: python
+        class Cluster(Compute):
+            def __cinit__(...):
+                self._set_compute_flag("computeClusters")
+
+            @Compute._compute("computeClusters")
+            def computeClusters(...)
+                ...
+
+            @Compute._computed_property("computeClusters")
+            def cluster_idx(self):
+                return ...
+
+            @Compute._reset
+            def reset(...):
+                ...
+
+    .. moduleauthor:: Jin Soo Ihm <jinihm@umich.edu>
+
+    Attributes:
+        _called_compute (dict):
+            Flags representing whether appropriate compute method was called.
+    """
+
+    def __cinit__(self):
+        self._called_compute = {"compute": False}
+
+    def _set_compute_flag(self, flag_name):
+        self._called_compute[flag_name] = False
+
+    @staticmethod
+    def _compute(key="compute"):
+        R"""Decorator that sets compute flag to be true.
+
+        Args:
+            key (str): Name of compute flag.
+
+        Returns:
+            Decorator decorating appropriate compute method.
+        """
+
+        def _compute_with_key(func):
+            @wraps(func)
+            def wrapper(self, *args, **kwargs):
+                retval = func(self, *args, **kwargs)
+                self._called_compute[key] = True
+                return retval
+            return wrapper
+        return _compute_with_key
+
+    @staticmethod
+    def _computed_property(key="compute"):
+        R"""Decorator that makes a class method to be a property with limited access.
+
+        Args:
+            key (str): Name of compute flag.
+
+        Returns:
+            Decorator decorating appropriate property method.
+        """
+        if isinstance(key, str):
+            key = (key,)
+
+        def _computed_property_with_key(func):
+            @property
+            @wraps(func)
+            def wrapper(self, *args, **kwargs):
+                if not any(self._called_compute[k] for k in key):
```

Comment:
If you change this to `self._called_compute.get(k, False)` then it will automatically return `False` if the key isn't set - this would avoid the need to call `self._set_compute_flag("computeClusters")` and similar methods in the constructors of the non-standard classes. Making `_called_compute` into a [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) is another option.

I'd prefer to remove the `_set_compute_flag` if possible, after that change is made.